documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,683 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+
+By virtue of the `State` properties not updating until `IInputBackend.Update` is called, the states enumerated by the lists will not change until `Update` is called.
+
+`Update` will call `IInputBackend.Update` on each of the `Backends`, and retrieve the state for each device contained in each backend. As it is doing this, it updates the lists within the input context.
+
+After the lists have been updated, the `Update` method compares the old state and the new state for that device and raises events accordingly. If there is no old state available for a device, the `ConnectionChanged` event (read on!) within each of the custom lists is called with a value of `true` signifying that a device has connected. Likewise, if there is no new state available for a device we have old state for, the `ConnectionChanged` event will be called with a value of `false` signifying that a device has been disconnected.
+
+`Backends` is a mutable list of input backends. Changes to this list do not take effect until `Update` is called again. The `ConnectionChanged` rules above will still be respected e.g. when you remove a backend, all of its devices will have a disconnected event raised for them.
+
+`Devices` contains all devices reported by all backends, including devices that do not necessarily fit into one of our more specialized wrapper lists. This means that if a backend has a device type we do not recognise, it will be accessible via this list.
+
+## Custom List Types
+
+These are relatively simple, though note that we explicitly implement the `IReadOnlyList` indexer in favour of one that returns a `ref readonly` type and a corresponding custom enumerator.
+
+The actual type of the custom enumerator at this time the Silk.NET team would like to leave implementation defined, but with the following requirements:
+- The `Next` method must return a `ref readonly` type.
+- The type can be used as an operand in a `foreach` loop
+- While it may not necessarily implement `IEnumerator<T>`, it must implement the surface API as if it does where `T` is a `ref readonly` type (which is forbidden in the language today).  
+
+The Silk.NET team does **not** wish to reserve the right to breaking changes for custom enumerator types, so once this type has been defined by the implementation it will remain the same type throughout the entirety of 3.X's service life. 
+
+All other details about the nature of the `IReadOnlyList` implementation (i.e. backing buffer) are implementation-defined.
+
+```cs
+public partial class Mice : IReadOnlyList<MouseState>
+{
+    public ref readonly MouseState this[int index] { get; }
+    MouseState IReadOnlyList<MouseState>.this[int index] { get; }
+    
+    public <implementation defined> GetEnumerator();
+    IEnumerator<MouseState> IEnumerable<MouseState>.GetEnumerator();
+    
+    public event Action<ButtonEvent<MouseState, MouseButton>>? ButtonDown;
```

Comment:
I'm a little bit on the fence about these events. Does it really make sense to have events indicating button up / button down? Can't we somehow indicate the "state delta"? Indicating what changed in the type seems wrong.

Reply:
Would an average user care enough?